metadata/src/audio/item.rs

```diff
@@ -16,98 +17,255 @@ use librespot_core::{
+
+        covers.sort_by(|a, b| b.width.cmp(&a.width));
+
+        covers
+            .iter()
+            .filter_map(|cover| {
+                let cover_id = cover.id.to_string();
+
+                if !cover_id.is_empty() {
+                    Some(image_url.clone().replace("{file_id}", &cover_id))
                 } else {
-                    return Err(UnavailabilityReason::NotWhitelisted);
+                    None
                 }
-            }
+            })
+            .collect()
+    }
+
+    pub fn artist(&self) -> Vec<String> {
+        match self {
+            AudioItem::Track(t) => t
+                .artists
+                .iter()
+                .map(|a| a.name.clone())
+                .collect::<Vec<String>>(),
+            AudioItem::Episode(e) => vec![e.description.clone()],
+        }
+    }
 
-            if let Some(forbidden_countries) = &premium_restriction.countries_forbidden {
-                if forbidden_countries
-                    .iter()
-                    .any(|forbidden| country == forbidden)
-                {
-                    return Err(UnavailabilityReason::Blacklisted);
+    pub fn album(&self) -> String {
+        match self {
+            AudioItem::Track(t) => t.album.name.clone(),
+            AudioItem::Episode(e) => e.publish_time.unix_timestamp().to_string(),
+        }
+    }
+
+    pub fn album_artist(&self) -> Vec<String> {
+        match self {
+            AudioItem::Track(t) => t
+                .album
+                .artists
+                .iter()
+                .map(|a| a.name.clone())
+                .collect::<Vec<String>>(),
+            AudioItem::Episode(e) => vec![e.show.clone()],
+        }
+    }
+
+    pub fn duration(&self) -> i32 {
+        match self {
+            AudioItem::Track(t) => t.duration,
+            AudioItem::Episode(e) => e.duration,
+        }
+    }
+
+    pub fn is_explicit(&self) -> bool {
+        match self {
+            AudioItem::Track(t) => t.is_explicit,
+            AudioItem::Episode(e) => e.is_explicit,
+        }
+    }
+
+    pub fn files(&self) -> &AudioFiles {
+        match self {
+            AudioItem::Track(t) => &t.files,
+            AudioItem::Episode(e) => &e.audio,
+        }
+    }
+
+    pub fn language(&self) -> Vec<String> {
+        match self {
+            AudioItem::Track(t) => t.language_of_performance.clone(),
+            AudioItem::Episode(e) => vec![e.language.clone()],
+        }
+    }
+
+    pub fn number(&self) -> i32 {
+        match self {
+            AudioItem::Track(t) => t.number,
+            AudioItem::Episode(e) => e.number,
+        }
+    }
+
+    pub fn disc_number(&self) -> i32 {
+        match self {
+            AudioItem::Track(t) => t.disc_number,
+            AudioItem::Episode(_) => 0,
+        }
+    }
+
+    pub fn popularity(&self) -> i32 {
+        match self {
+            AudioItem::Track(t) => t.popularity,
+            AudioItem::Episode(_) => 0,
```

Comment:
Or return an `Option` with `None`?

Reply:
I could but I figured this would be cheaper then sprinkling Options everywhere. It would either be ignored or turned into a 0 any later if some wanted to use it for some kind of rating or comparison.

Reply:
That's true, OK!

Reply:
All common fields are what they are and all unique fields are now options.